### PR TITLE
[arangodb] Add ArangoDB Support

### PR DESCRIPTION
**Pull request by smallsmallc**

### CODE REVIEW

**Comment by kruthar** [comment]
Can you focus this PR on just ArangoDB support and pull all the MongoDB stuff into it's own PR?

File: arangodb/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+Reload bash and test mvn
+
+    bash
+    mvn -version
+
+### 3. Set Up YCSB
+
+Clone this YCSB source code:
+
+    git clone https://***/YCSB.git
```
Comment (busbey) [inline arangodb/README.md]: either update to point at the repo, or better just reference folks to download the latest binary tarball.

File: arangodb/README.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+## ArangoDB Configuration Parameters
+
+- `arangodb.ip`
+  - Default value is `localhost`
+
+- `arangodb.port`
+  - Default value is `8529`.
+  
+- `arangodb.waitForSync`
+  - Default value is `false`.
```
Comment (busbey) [inline arangodb/README.md]: this should default to true.

File: arangodb/conf/logback.xml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<configuration>
```
Comment (busbey) [inline arangodb/conf/logback.xml]: needs a license header.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 4]
```diff
@@ -0,0 +1,488 @@
+/*
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: needs a license header.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+ * Arango client binding for YCSB.
+ *
+ * By Yemin, Sui
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: please rely on the git author for attribution.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This's thread-safe.
+ * Assume the table is the same during whole procedure.
+ * The data will be transfered from ByteIterator into String before sent to DB.
+ * 
+ * @author ysui
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: please rely on git authorship instead of author tags.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+   * one DB instance per client thread.
+   * 
+   * Actually, one client process will share one DB instance here.(Coincide to
+   * mongoDB driver)
+   */
+  @Override
+  public void init() throws DBException {
+    INIT_COUNT.incrementAndGet();
+    synchronized (ArangoDBClient.class) {
+      if (arangoDriver != null || arangoConfigure != null) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should this be an and? what if creating the driver object throws a runtime exception?

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+      arangoDriver = new ArangoDriver(arangoConfigure);
+
+      // Init the database
+      if (dropDBBeforeRun) {
+        // Try delete first
+        try {
+          arangoDriver.deleteDatabase(databaseName);
+        } catch (ArangoException e) {
+          if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DATABASE_NOT_FOUND) {
+            logger.error("Failed to delete database: " + databaseName, e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+        if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DUPLICATE_NAME) {
+          logger.error("Failed to create database: " + databaseName, e);
+          System.exit(-1);
+        } else {
+          logger.info("DB already exists: {}", databaseName);
+        }
+      }
+      // Always set the default db
+      arangoDriver.setDefaultDatabase(databaseName);
+      logger.info("ArangoDB client connection created to " + ip + ":" + port);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholders instead of string concat.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+      
+      // Log the configuration
+      StringBuilder configStr = new StringBuilder("Arango Configuration: ");
+      configStr.append(String.format("dropDBBeforeRun: %s; ", dropDBBeforeRun));
+      configStr.append(String.format("address: %s:%s; ", ip, port));
+      configStr.append(String.format("databaseName: %s; ", databaseName));
+      configStr.append(String.format("waitForSync: %s; ", waitForSync));
+      configStr.append(String.format("transactionUpdate: %s; ", transactionUpdate));
+      logger.info(configStr.toString());
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of calling toString

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+      if (defaultCollection == null) {
+        try {
+          CollectionOptions options = new CollectionOptions();
+          options.setWaitForSync(waitForSync);
+          arangoDriver.createCollection(table, options);
+          logger.info("New collection created: {} waitForSync: {}", table, waitForSync);
+        } catch (ArangoException e) {
+          if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DUPLICATE_NAME) {
+            logger.error("Fail to create collection", e);
+            System.exit(-1);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should never call System.exit. please find a way to propagate the error.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 12]
```diff
@@ -0,0 +1,488 @@
...
+    }
+  }
+
+  /**
+   * Can't create automatically during inserting, so create one if necessary.
+   * @param table
+   */
+  private void createCollection(String table) {
+    synchronized(ArangoDBClient.class) {
+      if (defaultCollection == null) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: this isn't going to properly handle when the table name changes across calls. the defaultCollection should be something like the last table accessed and it should be per DB instance.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 13]
```diff
@@ -0,0 +1,488 @@
...
+   *      The record key of the record to insert.
+   * @param values
+   *      A HashMap of field/value pairs to insert in the record
+   * @return Zero on success, a non-zero error code on error. See the
+   *     {@link DB} class's description for a discussion of error codes.
+   */
+  @Override
+  public Status insert(String table, String key, HashMap<String, ByteIterator> values) {
+    try {
+      if (defaultCollection == null) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: this initial check for null is good to avoid the synchronization check across DB instances, but it's not needed once you make the variable an instance variable. Also, to do this null check properly you'd have to make the variable `volatile` to ensure the JVM properly handled the fact that you're looking at it here outside of the lock.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 14]
```diff
@@ -0,0 +1,488 @@
...
+        createCollection(table);
+      }
+      BaseDocument toInsert = new BaseDocument(key);
+      for (Map.Entry<String, ByteIterator> entry : values.entrySet()) {
+        toInsert.addAttribute(entry.getKey(), byteIteratorToString(entry.getValue()));
+      }
+      arangoDriver.createDocument(table, toInsert, true/*create collection if not exist*/,
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: if we can pass a flag to the driver to create the collection if it doesn't exist, why the createCollection method?

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 15]
```diff
@@ -0,0 +1,488 @@
...
+                                  waitForSync);
+      return Status.OK;
+    } catch (ArangoException e) {
+      if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED) {
+        logger.error("Fail to insert: " + table + " " + key,  e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 16]
```diff
@@ -0,0 +1,488 @@
...
+      } else {
+        logger.debug("Trying to create document with duplicate key: {} {}", table, key);
+        return Status.BAD_REQUEST;
+      }
+    }  catch (Exception e) {
+      logger.error("Exception while trying insert " + table + " " + key, e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 17]
```diff
@@ -0,0 +1,488 @@
...
+  public Status read(String table, String key, Set<String> fields, HashMap<String, ByteIterator> result) {
+    DocumentCursor<BaseDocument> cursor = null;
+    try {
+      DocumentEntity<BaseDocument> targetDoc = arangoDriver.getDocument(table, key, BaseDocument.class);
+      BaseDocument aDocument = targetDoc.getEntity();
+      this.fillMap(result, aDocument.getProperties(), fields);
+      return Status.OK;
+    } catch (ArangoException e) {
+      if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DOCUMENT_NOT_FOUND) {
+        logger.error("Fail to read: " + table + " " + key,  e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 18]
```diff
@@ -0,0 +1,488 @@
...
+      } else {
+        logger.debug("Trying to read document not exist: {} {}", table, key);
+        return Status.NOT_FOUND;
+      }
+    } catch (Exception e) {
+      logger.error("Exception while trying read " + table + " " + key, e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 19]
```diff
@@ -0,0 +1,488 @@
...
+      this.fillMap(result, aDocument.getProperties(), fields);
+      return Status.OK;
+    } catch (ArangoException e) {
+      if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DOCUMENT_NOT_FOUND) {
+        logger.error("Fail to read: " + table + " " + key,  e);
+      } else {
+        logger.debug("Trying to read document not exist: {} {}", table, key);
+        return Status.NOT_FOUND;
+      }
+    } catch (Exception e) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: Exception is pretty broad. should this be RuntimeException?

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 20]
```diff
@@ -0,0 +1,488 @@
...
+      logger.error("Exception while trying read " + table + " " + key, e);
+    } finally {
+      if (cursor != null) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: it looks like cursor is always null? can it just be removed?

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 21]
```diff
@@ -0,0 +1,488 @@
...
+                mapToJson(values), Boolean.toString(waitForSync).toLowerCase());
+        TransactionEntity transaction = arangoDriver.createTransaction(transactionAction);
+        transaction.addWriteCollection(table);
+        transaction.setParams(createDocumentHandle(table, key));
+        arangoDriver.executeTransaction(transaction);
+        return Status.OK;
+      }
+    } catch (ArangoException e) {
+      if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DOCUMENT_NOT_FOUND) {
+        logger.error("Fail to update: " + table + " " + key,  e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 22]
```diff
@@ -0,0 +1,488 @@
...
+      } else {
+        logger.debug("Trying to update document not exist: {} {}", table, key);
+        return Status.NOT_FOUND;
+      }
+    } catch (Exception e) {
+      logger.error("Exception while trying update " + table + " " + key, e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 23]
```diff
@@ -0,0 +1,488 @@
...
+        return Status.OK;
+      }
+    } catch (ArangoException e) {
+      if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DOCUMENT_NOT_FOUND) {
+        logger.error("Fail to update: " + table + " " + key,  e);
+      } else {
+        logger.debug("Trying to update document not exist: {} {}", table, key);
+        return Status.NOT_FOUND;
+      }
+    } catch (Exception e) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: exception is very broad. should this be RuntimeException?

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 24]
```diff
@@ -0,0 +1,488 @@
...
+   *     {@link DB} class's description for a discussion of error codes.
+   */
+  @Override
+  public Status delete(String table, String key) {
+    try {
+      arangoDriver.deleteDocument(table, key);
+      return Status.OK;
+    } catch (ArangoException e) {
+      if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DOCUMENT_NOT_FOUND) {
+        logger.error("Fail to delete: " + table + " " + key,  e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 25]
```diff
@@ -0,0 +1,488 @@
...
+      } else {
+        logger.debug("Trying to delete document not exist: {} {}", table, key);
+        return Status.NOT_FOUND;
+      }
+    } catch (Exception e) {
+      logger.error("Exception while trying delete " + table + " " + key, e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 26]
```diff
@@ -0,0 +1,488 @@
...
+      arangoDriver.deleteDocument(table, key);
+      return Status.OK;
+    } catch (ArangoException e) {
+      if (e.getErrorNumber() != ErrorNums.ERROR_ARANGO_DOCUMENT_NOT_FOUND) {
+        logger.error("Fail to delete: " + table + " " + key,  e);
+      } else {
+        logger.debug("Trying to delete document not exist: {} {}", table, key);
+        return Status.NOT_FOUND;
+      }
+    } catch (Exception e) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: Exception is very broad. should this be RuntimeException?

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 27]
```diff
@@ -0,0 +1,488 @@
...
+      Iterator<BaseDocument> iterator = cursor.entityIterator();
+      while (iterator.hasNext()) {
+        BaseDocument aDocument = iterator.next();
+        HashMap<String, ByteIterator> aMap = new HashMap<String, ByteIterator>(aDocument.getProperties().size());
+        this.fillMap(aMap, aDocument.getProperties());
+        result.add(aMap);
+      }
+      return Status.OK;
+    } catch (Exception e) {
+      logger.error("Exception while trying scan " + table + " " + startkey + " " + recordcount, e);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: should use a format placeholder instead of string concat

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 28]
```diff
@@ -0,0 +1,488 @@
...
+      cursor = arangoDriver.executeDocumentQuery(aqlQuery, bindVars, null, BaseDocument.class);
+      Iterator<BaseDocument> iterator = cursor.entityIterator();
+      while (iterator.hasNext()) {
+        BaseDocument aDocument = iterator.next();
+        HashMap<String, ByteIterator> aMap = new HashMap<String, ByteIterator>(aDocument.getProperties().size());
+        this.fillMap(aMap, aDocument.getProperties());
+        result.add(aMap);
+      }
+      return Status.OK;
+    } catch (Exception e) {
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: Exception is very broad. Should this be RuntimeException?

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 29]
```diff
@@ -0,0 +1,488 @@
...
+  }
+
+  private String createDocumentHandle(String collectionName, String documentKey) throws ArangoException {
+    validateCollectionName(collectionName);
+    return collectionName + "/" + documentKey;
+  }
+
+  private void validateCollectionName(String name) throws ArangoException {
+    if (name.indexOf('/') != -1) {
+      throw new ArangoException("does not allow '/' in name.");
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: nit (feel free to fix or disregard): This might give clearer error messages if you used IllegalArgumentException here rather than ArangoException.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 30]
```diff
@@ -0,0 +1,488 @@
...
+  private void fillMap(Map<String, ByteIterator> resultMap, Map<String, Object> properties, Set<String> fields) {
+    if (fields == null || fields.size() == 0) {
+      for (Map.Entry<String, Object> entry : properties.entrySet()) {
+        if (entry.getValue() instanceof String) {
+          resultMap.put(entry.getKey(),
+              stringToByteIterator((String)(entry.getValue())));
+        } else {
+          logger.error("Error! Not the format expected! Actually is {}",
+              entry.getValue().getClass().getName());
+          System.exit(-1);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: Should never call System.exit. Please find a way to propagate the error instead.

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 31]
```diff
@@ -0,0 +1,488 @@
...
+        }
+      }
+    } else {
+      for (String field : fields) {
+        if (properties.get(field) instanceof String) {
+          resultMap.put(field, stringToByteIterator((String)(properties.get(field))));
+        } else {
+          logger.error("Error! Not the format expected! Actually is {}",
+              properties.get(field).getClass().getName());
+          System.exit(-1);
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: Should never call System.exit. Please find a way to propagate the error instead.

File: bin/ycsb [hunk 32]
```diff
@@ -210,7 +211,7 @@ def get_classpath_from_maven(module):
               "Maven follows, try running "
               "'mvn -DskipTests package dependency:build=classpath' on your "
               "own and correct errors." + os.linesep + os.linesep + "mvn output:" + os.linesep
-              + err.output)
+              + str(err.output))
```
Comment (busbey) [inline bin/ycsb]: this looks unrelated, please break it into a different PR.

File: mongodb/README.md [hunk 33]
```diff
@@ -145,8 +145,9 @@ See the next section for the list of configuration parameters for MongoDB.
 
 For example:
 
-    ./bin/ycsb load mongodb-async -s -P workloads/workloada -p mongodb.url=mongodb://localhost:27017/ycsb?w=0
+    ./bin/ycsb load mongodb-async -s -P workloads/workloada -p mongodb.url=mongodb://localhost:27017/ycsb?w=1
```
Comment (busbey) [inline mongodb/README.md]: this looks unrelated, please break it into a different PR.

File: mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java [hunk 34]
```diff
@@ -96,6 +96,8 @@
   /** A singleton Mongo instance. */
   private static MongoClient mongoClient;
 
+  private static Boolean dropDBBeforeRun;
```
Comment (busbey) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: this looks unrelated, please break it into a different PR.

Comment (smallsmallc) [inline mongodb/src/main/java/com/yahoo/ycsb/db/MongoDbClient.java]: Thanks for all your comments, that very nice and helpful. I've modified code. Thanks!

File: arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java [hunk 35]
```diff
@@ -0,0 +1,467 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * ArangoDB binding for YCSB framework using the ArangoDB Inc. <a
+ * href="https://github.com/arangodb/arangodb-java-driver">driver</a>
+ * <p>
+ * See the <code>README.md</code> for configuration information.
+ * </p>
+ * 
+ * @author smallsmallc
```
Comment (busbey) [inline arangodb/src/main/java/com/yahoo/ycsb/db/ArangoDBClient.java]: please don't include @author tags, we rely on git commit information

File: bin/ycsb [hunk 36]
```diff
@@ -53,6 +53,7 @@ COMMANDS = {
 DATABASES = {
     "accumulo"     : "com.yahoo.ycsb.db.accumulo.AccumuloClient",
     "aerospike"    : "com.yahoo.ycsb.db.AerospikeClient",
+    "arangodb"     : "com.yahoo.ycsb.db.ArangoDBClient",
```
Comment (busbey) [inline bin/ycsb]: needs a corresponding entry in bin/binding.properties

**Comment by busbey** [comment]
looks good! sorry for the delay, just those two last comments and I think we're good to go.
